pkg/agent/client/dial.go

```diff
@@ -65,9 +74,20 @@ func DialServer(ctx context.Context, config DialServerConfig) (*grpc.ClientConn,
 		}
 	}
 
-	return grpc.DialContext(ctx, config.Address,
+	ctx, cancel := context.WithTimeout(ctx, _defaultDialTimeout)
+	defer cancel()
+
+	if config.dialContext == nil {
+		config.dialContext = grpc.DialContext
+	}
+	client, err := config.dialContext(ctx, config.Address,
 		grpc.WithBalancerName(roundrobin.Name), //nolint:staticcheck
 		grpc.FailOnNonTempDialError(true),
+		grpc.WithBlock(),
 		grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig)),
 	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to dial %v: %v", config.Address, err)
+	}
```

Comment:
I wonder if we can get better error messaging by checking for specific error classes? "context deadline exceeded" makes sense to developers used to dealing with golang context's but not to the average operator.

```suggestion
	switch {
	case err == nil:
	case errors.Is(err, context.Canceled):
		return nil, fmt.Errorf("failed to dial %s: canceled", config.Address)
	case errors.Is(err, context.DeadlineExceeded):
		return nil, fmt.Errorf("failed to dial %s: timed out", config.Address)
	default:
		return nil, fmt.Errorf("failed to dial %s: %v", config.Address, err)
	}
```